### PR TITLE
Add session property for nested subfield filters

### PR DESCRIPTION
**Pull request by bhhari**
```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/ListSelectiveStreamReader.java [hunk 1]
```diff
@@ -572,7 +572,14 @@ public Block getBlock(int[] positions, int positionCount)
             nextPosition = positions[positionIndex];
         }
 
-        return ArrayBlock.fromElementBlock(positionCount, Optional.ofNullable(nullsCopy), offsetsCopy, elementStreamReader.getBlock(elementOutputPositions, elementOutputPositionCount));
+        Block elementBlock;
```
Comment (yingsu00) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/ListSelectiveStreamReader.java]: This change doesn't seem belong to this PR?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java [hunk 2]
```diff
@@ -366,6 +367,11 @@ public HiveSessionProperties(HiveClientConfig hiveClientConfig, OrcFileWriterCon
                         "Experimental: enable complex filter pushdown",
                         hiveClientConfig.isPushdownFilterEnabled(),
                         false),
+                booleanProperty(
```
Comment (yingsu00) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java]: Usually I see new properties were added to the end of the list, not in the middle, so people can see the order they're added. I'm not sure if this is the case? cc @wenleix

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 3]
```diff
@@ -1725,8 +1726,15 @@ public ConnectorPushdownFilterResult pushdownFilter(ConnectorSession session, Co
...
+        ColumnExtractor columnExtractor;
+        if (isSubFieldFilterEnabled(session, tableHandle)) {
+            columnExtractor = new SubfieldExtractor(functionResolution, rowExpressionService.getExpressionOptimizer(), session).toColumnExtractor();
+        }
+        else {
+            columnExtractor = (rowExpression, domain) -> Optional.empty();
+        }
         ExtractionResult<Subfield> decomposedFilter = rowExpressionService.getDomainTranslator()
-                .fromPredicate(session, filter, new SubfieldExtractor(functionResolution, rowExpressionService.getExpressionOptimizer(), session).toColumnExtractor());
+                .fromPredicate(session, filter, columnExtractor);
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: nit: put this with the previous line

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java [hunk 4]
```diff
@@ -1351,4 +1352,17 @@ public HiveClientConfig setPushdownFilterEnabled(boolean pushdownFilterEnabled)
         this.pushdownFilterEnabled = pushdownFilterEnabled;
         return this;
     }
+
+    public boolean isSubfieldFilterEnabled()
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java]: Shouldn't this be called subfield pruning or extracting or something?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 5]
```diff
@@ -1879,6 +1887,18 @@ private boolean isPushdownFilterEnabled(ConnectorSession session, ConnectorTable
         return false;
     }
 
+    private boolean isSubFieldFilterEnabled(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        boolean isSubFieldFilterEnabled = HiveSessionProperties.isSubFieldFilterEnabled(session);
+        if (isSubFieldFilterEnabled) {
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: This part of logic is unnecessary. Just inline this function

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java [hunk 6]
```diff
@@ -1351,4 +1352,17 @@ public HiveClientConfig setPushdownFilterEnabled(boolean pushdownFilterEnabled)
...
+    {
+        return subfieldFilterEnabled;
+    }
+
+    @Config(("hive.subfield-filter-enabled"))
+    @ConfigDescription(("Experimental: enable filters on nested columns"))
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java]: There are extra parentheses of these two.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java [hunk 7]
```diff
@@ -86,6 +86,7 @@
     private static final String TEMPORARY_TABLE_COMPRESSION_CODEC = "temporary_table_compression_codec";
     public static final String PUSHDOWN_FILTER_ENABLED = "pushdown_filter_enabled";
     public static final String VIRTUAL_BUCKET_COUNT = "virtual_bucket_count";
+    public static final String SUBFIELD_FILTER_ENABLED = "subfield_filter_enabled";
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java]: put this closer to `PUSHDOWN_FILTER_ENABLED`

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java [hunk 8]
```diff
@@ -284,6 +284,9 @@ public void testMaps()
                 .replaceAll("test_maps", "lineitem_ex")
                 .replaceAll("cardinality", "array_length");
         try {
+            //filter on nested columns
+            assertQueryUsingH2Cte("SELECT * FROM test_maps WHERE map_keys[1] > 10 and map_keys[1] < 20", rewriter);
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: just realized this change. What is `H2Cte`?

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/ListSelectiveStreamReader.java (hunk 1 above)
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/ListSelectiveStreamReader.java]: Looks like an unrelated change

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java (hunk 4 above)
Comment (bhhari) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java]: this session property is for enabling filters on subfields (nested columns)

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java (hunk 8 above)
Comment (bhhari) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: the assertQueryUsingH2Cte is added as h2 query required some rewriting as it does not match the same query as in presto

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/ListSelectiveStreamReader.java (hunk 1 above)
Comment (bhhari) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/ListSelectiveStreamReader.java]: There are test failures without this change

Comment (bhhari) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/ListSelectiveStreamReader.java]: there will be test failures without this change

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java (hunk 2 above)
Comment (bhhari) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java]: I wanted to keep this property next to the other related property

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/ListSelectiveStreamReader.java (hunk 1 above)
Comment (highker) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/ListSelectiveStreamReader.java]: What are those failures? If the failure is not related to the session property, maybe we could separate them out into other commits?

**Comment by bhhari** [comment]
#13304

File: presto-hive/src/main/java/com/facebook/presto/hive/SubfieldExtractor.java [hunk 9]
```diff
@@ -44,32 +44,56 @@
...
     private final ExpressionOptimizer expressionOptimizer;
     private final ConnectorSession connectorSession;
+    private final boolean nestedColumnsFilterEnabled;
 
     public SubfieldExtractor(
             StandardFunctionResolution functionResolution,
             ExpressionOptimizer expressionOptimizer,
-            ConnectorSession connectorSession)
+            ConnectorSession connectorSession,
+            boolean nestedColumnsFilterEnabled)
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/SubfieldExtractor.java]: This doesn't look right. `nestedColumnsFilterEnabled` is already in session property. You can get it from `connectorSession` above rather than creating a new field for it.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 10]
```diff
@@ -1748,7 +1748,14 @@ public ConnectorPushdownFilterResult pushdownFilter(ConnectorSession session, Co
...
         ExtractionResult<Subfield> decomposedFilter = rowExpressionService.getDomainTranslator()
-                .fromPredicate(session, filter, new SubfieldExtractor(functionResolution, rowExpressionService.getExpressionOptimizer(), session).toColumnExtractor());
+                .fromPredicate(
+                        session,
+                        filter,
+                        new SubfieldExtractor(
+                                functionResolution,
+                                rowExpressionService.getExpressionOptimizer(),
+                                session,
+                                HiveSessionProperties.isNestedColumnsFilterEnabled(session)).toColumnExtractor());
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: Just realized `toColumnExtractor` returns generic type. Could you fix the signature to be
```java
public DomainTranslator.ColumnExtractor<Subfield> toColumnExtractor()

File: presto-hive/src/test/java/com/facebook/presto/hive/TestDomainTranslator.java [hunk 11]
```diff
@@ -91,7 +91,7 @@ public void setup()
         columnExtractor = new SubfieldExtractor(
                 new FunctionResolution(metadata.getFunctionManager()),
                 (rowExpression, level, session) -> rowExpression,
-                TestingSession.SESSION).toColumnExtractor();
+                TestingSession.SESSION, true).toColumnExtractor();
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestDomainTranslator.java]: Create your own session to make the property to be true.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/ListSelectiveStreamReader.java [hunk 12]
```diff
@@ -572,7 +572,14 @@ public Block getBlock(int[] positions, int positionCount)
...
+        if (elementOutputPositionCount == 0) {
+            elementBlock = outputType.getElementType().createBlockBuilder(null, 0).build();
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/ListSelectiveStreamReader.java]: Could you make a helper `private static Block createEmptyBlock(Type type)` and consolidate all other use cases in this file as well?

File: presto-hive/src/main/java/com/facebook/presto/hive/SubfieldExtractor.java (hunk 9 above)
Comment (bhhari) [reply presto-hive/src/main/java/com/facebook/presto/hive/SubfieldExtractor.java]: I had this approach as there were some test failures involving TestingSession which throws exception when accessing a property. TestSubFieldExtractor. Will refactor and make it work

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java (hunk 2 above)
Comment (wenleix) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveSessionProperties.java]: I see both cases happened (added to the end of list vs. added next to related properties). I don't have strong opinions on this.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestDomainTranslator.java [hunk 13]
```diff
@@ -91,7 +99,7 @@ public void setup()
         columnExtractor = new SubfieldExtractor(
                 new FunctionResolution(metadata.getFunctionManager()),
                 (rowExpression, level, session) -> rowExpression,
-                TestingSession.SESSION).toColumnExtractor();
+                new TestConnectorSession()).toColumnExtractor();
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestDomainTranslator.java]: Instead of creating a new session class, just use
```java
TestingConnectorSession session = new TestingConnectorSession(new HiveSessionProperties(
        new HiveClientConfig().setNestedColumnsFilterEnabled(true),
        new OrcFileWriterConfig(),
        new ParquetFileWriterConfig())
        .getSessionProperties());

File: presto-hive/src/main/java/com/facebook/presto/hive/SubfieldExtractor.java [hunk 14]
```diff
@@ -55,21 +55,39 @@ public SubfieldExtractor(
...
+            // For complex types support only Filter IS_NULL and IS_NOT_NULL
+            if (isComplexType(type) && !(domain.isOnlyNull() || (domain.getValues().isAll() && !domain.isNullAllowed()))) {
+                return Optional.empty();
             }
 
-            return extract(expression);
+            Optional<Subfield> subfield = extract(expression);
+            // If the expression involves array or map subscripts, it is considered only if allowed by nested_columns_filter_enabled.
+            if (hasSubscripts(subfield)) {
+                if (HiveSessionProperties.isNestedColumnsFilterEnabled(connectorSession)) {
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/SubfieldExtractor.java]: static import `isNestedColumnsFilterEnabled`
